File: internal/observability/errs/posthog.go
package errs

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"github.com/posthog/posthog-go"
)

// posthogSink reports to PostHog's error tracking, which is the default backend
// for a Warmbly instance: the project key an operator may already have set for
// product analytics is the same key error events are captured with, so there is
// one vendor and one place to look instead of two.
//
// Nothing here identifies a person. The distinct id is the service name, not a
// user, and person profiles are switched off per event, so an exception never
// creates or updates a profile for anybody.
type posthogSink struct {
	client     posthog.Client
	distinctID string
	// base is the service, environment and release every event carries.
	base posthog.Properties
	// stacks resolves the Go stack into PostHog frames. Held rather than
	// constructed per event so the in-app rule is decided once.
	stacks posthog.DefaultStackTraceExtractor
	// logger is the SDK's, reused here so a failing backend costs one log line
	// per process rather than one per event.
	logger *posthogLogger
	closed sync.Once
}

// stackSkip drops runtime.Callers, the extractor, the sink method and this
// package's two fan-out frames, so the top frame of a reported stack is the
// line that called CaptureException. Every public entry point is the same
// depth from here, which is why one constant covers them all.
const stackSkip = 5

func newPostHogSink(cfg Config) (sink, error) {
	disableGeoIP := true
	logger := &posthogLogger{}
	client, err := posthog.NewWithConfig(cfg.PostHogKey, posthog.Config{
		// Empty means the SDK's own default, PostHog Cloud US.
		Endpoint: cfg.PostHogHost,
		// A server's IP is the datacentre's, so geolocating it says nothing
		// and storing it is worse than useless.
		DisableGeoIP: &disableGeoIP,
		// Close waits forever without this, so an unreachable host would hang
		// the exit path that flushes rather than delay it.
		ShutdownTimeout: fatalFlushTimeout,
		Logger:          logger,
	})
	if err != nil {
		return nil, fmt.Errorf("init posthog error tracking: %w", err)
	}

	base := posthog.Properties{"service": cfg.Service}
	if cfg.Environment != "" {
		base["environment"] = cfg.Environment
	}
	if cfg.Release != "" {
		base["release"] = cfg.Release
	}

	return &posthogSink{
		client:     client,
		distinctID: distinctID(cfg.Service),
		base:       base,
		stacks:     posthog.DefaultStackTraceExtractor{InAppDecider: posthog.SimpleInAppDecider},
		logger:     logger,
	}, nil
}

func (p *posthogSink) capture(ev event) {
	stack := p.stacks.GetStackTrace(stackSkip)
	handled := true

	if ev.err != nil {
		p.enqueue(ev.scope, "error", posthog.ExceptionItem{
			// The Go type is the issue title, the message its description.
			Type:       fmt.Sprintf("%T", ev.err),
			Value:      ev.err.Error(),
			Mechanism:  &posthog.ExceptionMechanism{Handled: &handled},
			Stacktrace: stack,
		})
		return
	}

	p.enqueue(ev.scope, "info", posthog.ExceptionItem{
		// A fixed title, because a message's text is the description and
		// putting it in the title would make one issue per wording.
		Type:       "message",
		Value:      ev.message,
		Mechanism:  &posthog.ExceptionMechanism{Handled: &handled},
		Stacktrace: stack,
	})
}

func (p *posthogSink) recoverPanic(_ context.Context, r any, sc scope) {
	stack := p.stacks.GetStackTrace(stackSkip)
	handled := false

	p.enqueue(sc, "error", posthog.ExceptionItem{
		Type:       "panic",
		Value:      fmt.Sprint(r),
		Mechanism:  &posthog.ExceptionMechanism{Handled: &handled},
		Stacktrace: stack,
	})
}

// flush closes the client, which is the only flush the SDK offers. Callers are
// on their way out of the process, which is why that is acceptable here and why
// Flush is documented as an exit-path call.
func (p *posthogSink) flush(time.Duration) bool {
	drained := false
	p.closed.Do(func() {
		drained = p.client.Close() == nil
	})
	return drained
}

func (p *posthogSink) enqueue(sc scope, level string, item posthog.ExceptionItem) {
	props := posthog.Properties{
		// No person is created or updated by an exception: the distinct id
		// names a process, not somebody.
		"$process_person_profile": false,
		"$exception_level":        level,
	}
	for k, v := range p.base {
		props[k] = v
	}
	for k, v := range sc.tags {
		props[k] = v
	}
	for k, v := range sc.extra {
		props[k] = v
	}

	err := p.client.Enqueue(posthog.Exception{
		DistinctId:    p.distinctID,
		Timestamp:     time.Now().UTC(),
		Properties:    props,
		ExceptionList: []posthog.ExceptionItem{item},
	})
	if err != nil {
		p.logger.Errorf("cannot queue an exception: %v", err)
	}
}

// distinctID names the process rather than a person. PostHog requires one on
// every event, so it gets the least identifying value that still separates the
// backend's issues from a worker's.
func distinctID(service string) string {
	if service == "" {
		return "warmbly"
	}
	return "warmbly-" + service
}

// posthogLogger keeps the SDK's own chatter out of the instance's logs and
// bounds its failures to one line per process.
//
// A wrong key or an unreachable host fails on every single batch, so logging
// each one would bury the real logs under reporting noise. Logging none of them
// is worse: a misconfigured key would look exactly like a quiet week.
type posthogLogger struct {
	warned sync.Once
}

func (l *posthogLogger) Debugf(string, ...any) {}
func (l *posthogLogger) Logf(string, ...any)   {}

func (l *posthogLogger) Warnf(format string, args ...any) {
	l.Errorf(format, args...)
}

func (l *posthogLogger) Errorf(format string, args ...any) {
	l.warned.Do(func() {
		log.Printf("error reporting to PostHog is failing for this run (check POSTHOG_KEY and POSTHOG_HOST): "+format, args...)
	})
}

File: tracking/src/posthog.rs
//! PostHog error tracking for the tracking service.
//!
//! There is no PostHog crate in this build, and there does not need to be: an
//! `$exception` event is one JSON POST to the same capture endpoint the
//! backend's analytics already uses, and the payload shape is the one PostHog
//! documents for manual capture. That keeps the pixel's dependency list, and so
//! its image, unchanged.
//!
//! Sending happens on one dedicated thread with a blocking client rather than
//! spawned onto the request runtime, so a slow or unreachable host can never
//! delay a pixel, and so `flush` on a fatal path has something to wait on.

use serde_json::{json, Value};
use std::sync::mpsc::{sync_channel, Receiver, SyncSender};
use std::sync::OnceLock;
use std::time::Duration;
use tracing::warn;

/// PostHog Cloud US, where most of the customers are.
const DEFAULT_HOST: &str = "https://us.i.posthog.com";

/// PostHog's single-event capture endpoint.
const CAPTURE_PATH: &str = "/i/v0/e/";

/// One send. Long enough for a slow host, short enough that a stuck queue
/// drains rather than grows.
const SEND_TIMEOUT: Duration = Duration::from_secs(5);

/// Queue depth. A service reporting faster than this drops rather than grows a
/// backlog: the newest errors are worth no more than the pixel's memory.
const QUEUE_DEPTH: usize = 64;

static REPORTER: OnceLock<Reporter> = OnceLock::new();

enum Message {
    Event(Value),
    /// Acked once every event queued ahead of it has been sent.
    Drain(SyncSender<()>),
}

struct Reporter {
    tx: SyncSender<Message>,
    key: String,
    context: Value,
}

/// Starts the sender thread. Returns false when no key is configured, which is
/// the default and means nothing is ever sent.
pub fn init(key: &str, host: &str, env: &str, release: &str) -> bool {
    let key = key.trim();
    if key.is_empty() {
        return false;
    }

    let host = match host.trim().trim_end_matches('/') {
        "" => DEFAULT_HOST.to_string(),
        h => h.to_string(),
    };

    let (tx, rx) = sync_channel(QUEUE_DEPTH);
    std::thread::Builder::new()
        .name("posthog-errors".into())
        .spawn(move || send_loop(host, rx))
        .ok();

    REPORTER
        .set(Reporter {
            tx,
            key: key.to_string(),
            context: json!({
                "service": "tracking",
                "environment": env,
                "release": release,
            }),
        })
        .is_ok()
}

/// Queues one exception. Never blocks: a full queue drops the event.
pub fn capture(exception_type: &str, value: &str, handled: bool) {
    let Some(reporter) = REPORTER.get() else {
        return;
    };

    let mut properties = json!({
        "$exception_list": [{
            "type": exception_type,
            "value": value,
            "mechanism": { "handled": handled, "synthetic": false },
        }],
        "$exception_level": "error",
        // No person is created or updated by an exception: the distinct id
        // names a process, not somebody.
        "$process_person_profile": false,
        // A server's IP is the datacentre's, so geolocating it says nothing.
        "$geoip_disable": true,
    });
    if let (Some(props), Some(context)) = (properties.as_object_mut(), reporter.context.as_object())
    {
        for (k, v) in context {
            props.insert(k.clone(), v.clone());
        }
    }

    let event = json!({
        "api_key": reporter.key,
        "event": "$exception",
        "distinct_id": "warmbly-tracking",
        "properties": properties,
        "timestamp": chrono::Utc::now().to_rfc3339(),
    });

    let _ = reporter.tx.try_send(Message::Event(event));
}

/// Waits up to `timeout` for the queue to drain. `std::process::exit` skips
/// every destructor, so without this the error a fatal path just reported is
/// exactly the one that never arrives.
pub fn flush(timeout: Duration) {
    let Some(reporter) = REPORTER.get() else {
        return;
    };
    let (ack, done) = sync_channel(1);
    // try_send, not send: a full queue against a dead host would block the exit
    // path for as long as it takes to time out every event ahead of us, and a
    // queue that full is not going to drain inside `timeout` anyway.
    if reporter.tx.try_send(Message::Drain(ack)).is_ok() {
        let _ = done.recv_timeout(timeout);
    }
}

fn send_loop(host: String, rx: Receiver<Message>) {
    let client = match reqwest::blocking::Client::builder()
        .timeout(SEND_TIMEOUT)
        .build()
    {
        Ok(c) => c,
        Err(e) => {
            warn!("PostHog error tracking disabled, no HTTP client: {e}");
            return;
        }
    };
    let url = format!("{host}{CAPTURE_PATH}");
    // A wrong key fails on every event, so the failure is logged once rather
    // than once per error; logging none of them would make a misconfigured key
    // look exactly like a quiet week.
    let mut warned = false;

    while let Ok(message) = rx.recv() {
        match message {
            Message::Drain(ack) => {
                let _ = ack.try_send(());
            }
            Message::Event(event) => {
                let result = client.post(&url).json(&event).send();
                if warned {
                    continue;
                }
                match result {
                    Ok(response) if !response.status().is_success() => {
                        warn!(
                            "PostHog rejected an error report with {} (check POSTHOG_KEY)",
                            response.status()
                        );
                        warned = true;
                    }
                    Err(e) => {
                        warn!("cannot reach the PostHog host {url}: {e}");
                        warned = true;
                    }
                    _ => {}
                }
            }
        }
    }
}
